[MIG] stock_manual_transfer: Migration to 16.0 T#81066

allanpa88 opened the pull request:
- Changed name for deprecated modules like stock.location.route to
  stock.route and picking.moves to move.line.ids.
- Added user to group uom to fix unit tests and avoid error for not
  null violation in product uom.
- Improved preferred Route domain to show only stock routes related
  to the supplied warehouse.
- Added return early pattern in action_validate() method to avoid
  duplicated transfers after first validation and state change.

Commit:
[ADD] stock_manual_transfer: New module added

- New module added in order to allow the user to create stock transfers
  manually.

carralc committed:
[FIX] stock_manual_transfer: remove .unlink() method override

Due to upstream updates adding support for @api.ondelete and
pre-commit-vauxoo detecting the override of .unlink() method as
a code smell, this transforms the old logic into more
idiomatic odoo code.

carralc committed:
[FIX] stock_manual_transfer: super().unlink() was not removed in original commit

While the original fix (https://github.com/Vauxoo/addons-vauxoo/commit/cbd70ffe5b6693e9ecd299e3c935a3dcfa4a0131)
did fix the immediate lint messages, we didn't actually stop it from
calling super().unlink(), thus it didn't actually fix what it intended
to fix. Here we remove that call to super().unlink() so the issue is
properly fixed.

allanpa88 committed:
[MIG] stock_manual_transfer: Migration to 16.0 #81066

- Changed name for deprecated models like stock.location.route to
  stock.route and stock move to Operations(stock.move.line) to Stock
  Moves.
- Added user to group uom to fix unit tests and avoid error for not
  null violation in product uom.

hugho-ad commented on stock_manual_transfer/models/stock_manual_transfer.py:
```diff
@@ -0,0 +1,121 @@
+    )
+
+    def action_validate(self):
+        if self.state == "valid":
```
```suggestion
        if self.state != "draft":
```

allanpa88 replied:
Done!

hugho-ad reviewed (commented):
Just one comment, then LGTM :+1:

luisg123v commented on stock_manual_transfer/demo/res_groups_demo.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
```
Demo data shouldn't be marked as noupdate.

luisg123v commented on stock_manual_transfer/demo/res_groups_demo.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
+    <record model="res.groups" id="uom.group_uom">
```
- ID goes before model
- Modify the user, not the group
- I don't like this, this is a technical group, which mean it should be enabled or disabled for all users, but you're enabling it for a single user.

allanpa88 replied:
This is not required, so I will delete the file. However, manual access will be needed for testing purposes.

The original idea was to create a plug-and-play environment where you could start testing functionality without the need for prior configurations.

luisg123v commented on stock_manual_transfer/models/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from . import procurement_group
+from . import stock_location
```
`stock_route`

luisg123v commented on stock_manual_transfer/models/stock_location.py:
```diff
@@ -0,0 +1,7 @@
+from odoo import fields, models
+
+
+class StockLocationRoute(models.Model):
```
Rename also class.

luisg123v commented on stock_manual_transfer/tests/test_stock_manual_transfer.py:
```diff
@@ -0,0 +1,89 @@
+        return transfer
+
+    def create_manual_transfer_line(self, transfer):
+        self.env.ref("uom.group_uom").write({"users": [(4, self.env.user.id)]})
```
- Not correct, this is not part of a manual transfer creation. If really necessary, this should be in the SetUpClass method.
- Use the command syntax

allanpa88 replied:
It's required to avoid a not null Violation error with the product uom relation in the transfer lines.

luisg123v commented on stock_manual_transfer/tests/test_stock_manual_transfer.py:
```diff
@@ -26,9 +26,9 @@ def create_manual_transfer(self):
+        self.env.ref("uom.group_uom").write({"users": [(4, self.env.user.id)]})
         with Form(transfer) as tr, tr.transfer_line_ids.new() as line:
             line.product_id = self.product
-            line.product_uom_qty = 1
```
Why?

allanpa88 replied:
It is not needed at this point in the test, as the value is already set to 1. However, we can keep the line even though it is unnecessary.

luisg123v commented on stock_manual_transfer/demo/res_groups_demo.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo noupdate="1">
+    <record model="res.groups" id="uom.group_uom">
+        <field name="users" eval="[(4, ref('base.user_admin'))]" />
```
Use the Command syntax

luisg123v reviewed (changes requested):
Document performed changes in commit description.

luisg123v commented:
@allanpa88 Please check the migration done by @xmglord on #1649

FYI @xmglord we will base the migration to 17.0 on top of this one.

luisg123v commented on stock_manual_transfer/__manifest__.py:
```diff
@@ -0,0 +1,29 @@
+        "security/res_groups.xml",
+        "security/ir.model.access.csv",
+        "data/ir_sequence_data.xml",
+        "views/stock_location_route_views.xml",
```
Rename file.

luisg123v commented on stock_manual_transfer/models/stock_manual_transfer.py:
```diff
@@ -63,6 +63,8 @@ class StockManualTransfer(models.Model):
 
     def action_validate(self):
+        if self.state != "draft":
+            return
```
How is this possible?

allanpa88 replied:
"I don't have an idea how to reproduce the error. But is an issue reported.
 Please remember that this module is already in use but was installed from the Client Road main app (https://git.vauxoo.com/vauxoo/7llantas)
 
 The PM's group has declared it as an issue that cannot be replicated, but a couple of times, the transfers originated from this module has been found duplicated.

For more context, please refer to the comments by Julio Zerna and Zuly Lira in helpdesk ticket 81066. They are the authors of the return early suggestion as an answer of this issue

https://www.vauxoo.com/web#id=81066&cids=1&menu_id=1431&model=project.task&view_type=form

luisg123v commented on stock_manual_transfer/models/stock_manual_transfer.py:
```diff
@@ -28,9 +28,9 @@ class StockManualTransfer(models.Model):
+        "stock.route",
         string="Preferred Route",
-        domain=[("manual_transfer_selectable", "=", True)],
+        domain="[('manual_transfer_selectable', '=', True), ('supplied_wh_id', '=', warehouse_id)]",
```
Are you sure? Take into account warehouse fields are not mandatory in routes.

allanpa88 replied:
Yes, I'm experiencing the same issue. 
This feature was requested by Zuly Lira and is already in the production environment. 
When I asked her if this feature was required for all customers, she confirmed that it was a better way to work with the module. This is the origin for that change:
https://git.vauxoo.com/vauxoo/7llantas/-/commit/94a57eca77c58724ef2f3412a0fd921e39d0a420

Here is a snippet of our conversation and her request: 
Zuly Lira, [9/5/24 11:19 A. M.]
Hola @Allan Porras [Vauxoo] recuerdo que hicimos una mejora para que en rutas solo mostrar las relacionadas al almacén seleccionado https://imgur.com/Vq7Mufb y haciendo pruebas ahora en el staging, aparecen todas las rutas https://imgur.com/45Zd5vx
Me ayudas a revisar por favor.
Gracias.

allanpa88, [9/5/24 11:24 A. M.]
Revisando
Igual ese módulo como te había mencionado se va a pasar a addons-vauxoo
o sea, ya no es sólo de Road

allanpa88, [9/5/24 11:25 A. M.]
consideras que es una característica que se deba mantener en el módulo general?

Zuly Lira, [9/5/24 11:25 A. M.]
siiii
es mucho mas funcional así...

And it's true, if you select an incorrect route of the entire list, you could get some functional errors difficult to resolve for a non technical or Odoo expert user.

luisg123v reviewed (changes requested):
Commit message:
- "Changed name for deprecated modules like stock.location.route": not a module but a model
- "picking.moves to move.line.ids": these models don't exist

allanpa88 committed:
[IMP] stock_manual_transfer:  Add domain to route_id field

Add domain to route_id field to show the only the routes that are
supplied_wh_id = warehouse_id. This is to avoid
inconsistencies between the rule and the warehouse.

allanpa88 committed:
[FIX] stock_manual_transfer: Prevent creation of duplicate transfers

To prevent any potential duplication, this fix introduces a
return-early pattern. Now, transfers are not created if the manual
transfer record has already been published.

Issue reported on helpdesk ticket #25879

luisg123v reviewed (approved):
LGTM :+1:

xmglord commented on stock_manual_transfer/tests/test_stock_manual_transfer.py:
```diff
@@ -8,6 +8,7 @@ class TestStock(TransactionCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
+        cls.env.user.groups_id += cls.env.ref("uom.group_uom")
```
@luisg123v this shouldn't be done, there is a group restriction in the stock manual transfer view over the unit of measure in [here](https://github.com/Vauxoo/addons-vauxoo/blob/16.0/stock_manual_transfer/views/stock_manual_transfer_views.xml#L47) and in v15 it worked fine, the field didn't show and a user without the multi unit of measure group and they can make the manual transfers but now in the v16 you cannot preform a manual transfer if you don't belong to the multi unit group.

![image](https://github.com/Vauxoo/addons-vauxoo/assets/119946425/36d885ab-b64b-42d5-a3e0-e6086125e166)

cc: @allanpa88
